### PR TITLE
Core: Add metrics reporter for serializable table

### PR DESCRIPTION
**Pull request by wangyinsheng**
The metrics reporter will be lost when a table be sent to other nodes in a cluster through serializable table, This mr rebuild  the metrics reporter when the table is deserialized

### CODE REVIEW

**Comment by wangyinsheng** [comment]
PTAL @nastra

File: core/src/main/java/org/apache/iceberg/SerializableTable.java [hunk 1]
```diff
@@ -247,6 +252,18 @@ public Map<String, SnapshotRef> refs() {
     return refs;
   }
 
+  @Override
+  public MetricsReporter metricsReporter() {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/SerializableTable.java:256]: can we add a test that makes sure the metrics reporter is preserved?

Comment (wangyinsheng) [reply core/src/main/java/org/apache/iceberg/SerializableTable.java:256]: PTAL

File: api/src/test/java/org/apache/iceberg/TestHelpers.java [hunk 2]
```diff
@@ -137,6 +142,18 @@ public static void assertSerializedAndLoadedMetadata(Table expected, Table actua
     Assert.assertEquals("History must match", expected.history(), actual.history());
   }
 
+  public static void assertSerializedMetricsReporter(
```
Comment (nastra) [inline api/src/test/java/org/apache/iceberg/TestHelpers.java:145]: no need to add this here, this can be just part of the test method itself

File: api/src/test/java/org/apache/iceberg/TestHelpers.java [hunk 3]
```diff
@@ -137,6 +142,18 @@ public static void assertSerializedAndLoadedMetadata(Table expected, Table actua
...
+      MetricsReporter expected, MetricsReporter actual) {
+    Assert.assertNotNull("metrics reporter from serializableTable should not be null", actual);
+    Assert.assertTrue(
+        "metrics reporter from serializableTable should be instance of TestMetricsReporter",
+        actual instanceof TestMetricsReporter);
+    Assert.assertEquals(
+        "metrics reporter from serializableTable should equals the one from origin table",
+        expected,
+        actual);
```
Comment (nastra) [inline api/src/test/java/org/apache/iceberg/TestHelpers.java:154]: rather than testing actual equality on the objects, it should be enough to make sure that the properties are the same, hence the `TestMetricsReporter` wouldn't need to implement hasCode/equals

File: core/src/test/java/org/apache/iceberg/TestSerializableTable.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.io.IOException;
+import java.util.Map;
+import org.apache.iceberg.metrics.MetricsReporter;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.Types;
+import org.junit.After;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestSerializableTable.java:34]: please take a look at https://iceberg.apache.org/contribute/#testing. Basically new test classes should be JUnit5 rather than JUnit4

File: core/src/test/java/org/apache/iceberg/TestTables.java [hunk 5]
```diff
@@ -264,8 +265,17 @@ public void commit(TableMetadata base, TableMetadata updatedMetadata) {
             throw new CommitFailedException("Injected failure");
           }
           Integer version = VERSIONS.get(tableName);
+          String metadataLocation =
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestTables.java:268]: why are these changes necessary?

File: api/src/test/java/org/apache/iceberg/TestHelpers.java [hunk 6]
```diff
@@ -137,6 +142,18 @@ public static void assertSerializedAndLoadedMetadata(Table expected, Table actua
     Assert.assertEquals("History must match", expected.history(), actual.history());
   }
 
+  public static void assertSerializedMetricsReporter(
+      MetricsReporter expected, MetricsReporter actual) {
+    Assert.assertNotNull("metrics reporter from serializableTable should not be null", actual);
+    Assert.assertTrue(
```
Comment (nastra) [inline api/src/test/java/org/apache/iceberg/TestHelpers.java:148]: this can be simplified to `Assertions.assertThat(actual).isNotNull().isInstanceOf(...)` and for the properties: `Assertions.assertThat(actual.properties).isEqualTo(...)`

**Review by nastra — CHANGES_REQUESTED** [review]

File: api/src/test/java/org/apache/iceberg/TestHelpers.java (hunk 2 above)
Comment (wangyinsheng) [reply api/src/test/java/org/apache/iceberg/TestHelpers.java]: done

File: api/src/test/java/org/apache/iceberg/TestHelpers.java (hunk 3 above)
Comment (wangyinsheng) [reply api/src/test/java/org/apache/iceberg/TestHelpers.java]: done

File: core/src/test/java/org/apache/iceberg/TestSerializableTable.java (hunk 4 above)
Comment (wangyinsheng) [reply core/src/test/java/org/apache/iceberg/TestSerializableTable.java]: PTAL

File: api/src/test/java/org/apache/iceberg/TestHelpers.java (hunk 6 above)
Comment (wangyinsheng) [reply api/src/test/java/org/apache/iceberg/TestHelpers.java]: done, thanks

File: core/src/test/java/org/apache/iceberg/TestTables.java (hunk 5 above)
Comment (wangyinsheng) [reply core/src/test/java/org/apache/iceberg/TestTables.java:268]: the method `lazyTable` in `SerializableTable` will load metadata file,  while the `TestTable` has not write metadata file

File: core/src/test/java/org/apache/iceberg/TestSerializableTable.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  @AfterAll
+  public static void clean() {
+    TestTables.clearTables();
+  }
+
+  @Test
+  public void testSerializableTableWithMetricsReporter()
+      throws IOException, ClassNotFoundException {
+    Map<String, String> properties = Maps.newHashMap();
+    properties.put(
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestSerializableTable.java:61]: nit: can be simplified to `Map<String, String> properties = ImmutableMap.of(...)`

File: core/src/test/java/org/apache/iceberg/TestSerializableTable.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+          optional(2, "data", Types.StringType.get()),
+          required(3, "date", Types.StringType.get()),
+          optional(4, "double", Types.DoubleType.get()));
+
+  private static final PartitionSpec SPEC =
+      PartitionSpec.builderFor(SCHEMA).identity("date").build();
+
+  private static final SortOrder SORT_ORDER = SortOrder.builderFor(SCHEMA).asc("id").build();
+
+  @TempDir public File temp;
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestSerializableTable.java:50]: I think this doesn't have to be public to work

File: api/src/main/java/org/apache/iceberg/Table.java [hunk 9]
```diff
@@ -346,4 +347,14 @@ default Snapshot snapshot(String name) {
...
   }
+
+  /**
+   * Returns the metrics reporter for this table.
+   *
+   * @return the metrics reporter for this table.
+   */
+  default MetricsReporter metricsReporter() {
+    throw new UnsupportedOperationException(
+        "metricsReporter is not supported by " + getClass().getName());
```
Comment (nastra) [inline api/src/main/java/org/apache/iceberg/Table.java:358]:
```suggestion
        "Accessing metrics reporter is not supported");
```

File: api/src/test/java/org/apache/iceberg/TestHelpers.java [hunk 10]
```diff
@@ -559,4 +561,21 @@ public List<Long> splitOffsets() {
       return null;
     }
   }
+
+  public static class TestMetricsReporter implements MetricsReporter {
```
Comment (nastra) [inline api/src/test/java/org/apache/iceberg/TestHelpers.java:565]: can we also move this to the respective test class? I don't think it's necessary to put this into `TestHelpers`

File: core/src/test/java/org/apache/iceberg/TestSerializableTable.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  @Test
+  public void testSerializableTableWithMetricsReporter()
+      throws IOException, ClassNotFoundException {
+    Map<String, String> properties = Maps.newHashMap();
+    properties.put(
+        CatalogProperties.METRICS_REPORTER_IMPL, TestHelpers.TestMetricsReporter.class.getName());
+    MetricsReporter reporter = CatalogUtil.loadMetricsReporter(properties);
+    Table table = TestTables.create(temp, "tbl_A", SCHEMA, SPEC, SORT_ORDER, 2, reporter);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestSerializableTable.java:64]: I don't see where we're testing that the reporter itself has the correct properties.

Shouldn't this do
```
Map<String, String> reporterProperties = ImmutableMap.of("a", "1", "b", "2");
reporter.initialize(reporterProperties);
```

File: core/src/main/java/org/apache/iceberg/SerializableTable.java [hunk 12]
```diff
@@ -247,6 +252,18 @@ public Map<String, SnapshotRef> refs() {
...
+    if (lazyMetricsReporter == null) {
+      synchronized (this) {
+        if (lazyMetricsReporter == null) {
+          lazyMetricsReporter = CatalogUtil.loadMetricsReporter(this.metricsReporterProperties);
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/SerializableTable.java:260]: this is actually wrong and this should pass `properties` so that we have `CatalogUtil.loadMetricsReporter(this.properties)`

To finish the initialization of the metrics reporter after it was loaded, you'd have to do `lazyMetricsReporter.initialize(this.metricsReporterProperties)`

File: core/src/test/java/org/apache/iceberg/TestSerializableTable.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+  @AfterAll
+  public static void clean() {
+    TestTables.clearTables();
+  }
+
+  @Test
+  public void testSerializableTableWithMetricsReporter()
+      throws IOException, ClassNotFoundException {
+    Map<String, String> properties = Maps.newHashMap();
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestSerializableTable.java:60]: those are the properties of the table, so you'd have to pass them to the table. The metrics reporter then has separate properties

**Review by nastra — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/apache/iceberg/TestSerializableTable.java (hunk 7 above)
Comment (wangyinsheng) [reply core/src/test/java/org/apache/iceberg/TestSerializableTable.java]: done

File: core/src/test/java/org/apache/iceberg/TestSerializableTable.java (hunk 8 above)
Comment (wangyinsheng) [reply core/src/test/java/org/apache/iceberg/TestSerializableTable.java]: done

File: api/src/main/java/org/apache/iceberg/Table.java (hunk 9 above)
Comment (wangyinsheng) [reply api/src/main/java/org/apache/iceberg/Table.java]: done

File: api/src/test/java/org/apache/iceberg/TestHelpers.java (hunk 10 above)
Comment (wangyinsheng) [reply api/src/test/java/org/apache/iceberg/TestHelpers.java]: done

File: core/src/test/java/org/apache/iceberg/TestSerializableTable.java (hunk 11 above)
Comment (wangyinsheng) [reply core/src/test/java/org/apache/iceberg/TestSerializableTable.java:68]: `reporter.initialize(reporterProperties)` is called in the method `CatalogUtil.loadMetricsReporter(properties)`

File: core/src/main/java/org/apache/iceberg/SerializableTable.java (hunk 12 above)
Comment (wangyinsheng) [reply core/src/main/java/org/apache/iceberg/SerializableTable.java:260]: metrics reporter is configured by catalog properties, not table properties. I add the interface method `MetricsReporter.properties()` to return the  properties which used to load and initialize this `MetricsReporter`,  therefor the `SerializableTable` can use this properties to load and initialize `MetricsReporter` agagin

File: core/src/test/java/org/apache/iceberg/TestSerializableTable.java (hunk 13 above)
Comment (wangyinsheng) [reply core/src/test/java/org/apache/iceberg/TestSerializableTable.java]: metrics reporter is configured by catalog properties, not table properties. Here this `properties`  is taken as properties of catalog

File: core/src/main/java/org/apache/iceberg/SerializableTable.java (hunk 12 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/SerializableTable.java:260]: sorry my bad, you're right. it should be the catalog properties

File: core/src/test/java/org/apache/iceberg/TestTables.java (hunk 5 above)
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/TestTables.java:268]: I still think those changes are not necessary. The test passed fine for me without having those changes

**Review by nastra — COMMENTED** [review]
I did take another look and I don't think we need the changes in `TestTables`, because the tests will pass without those.

**Comment by wangyinsheng** [comment]
> I did take another look and I don't think we need the changes in `TestTables`, because the tests will pass without those.

you're right, it not necessary. I used to compare the metadata between origin table and serializable table, then it will be needed. I have removed the changes, PTAL @nastra

**Review by nastra — APPROVED** [review]
LGTM, thanks @wangyinsheng.

@aokolnychyi could you also take a look please?

**Comment by wangyinsheng** [comment]
PTAL @aokolnychyi

**Comment by wangyinsheng** [comment]
PTAL @aokolnychyi

**Comment by wangyinsheng** [comment]
could you please take a look this pr, it has last for weeks @aokolnychyi

**Comment by wangyinsheng** [comment]
@nastra  Could you please approval the CI ?  And is there someone else can review this pr?

**Comment by wangyinsheng** [comment]
@aokolnychyi @danielcweeks could you take a look please ? thanks

**Comment by nastra** [comment]
@wangyinsheng to fix CI failures I think you need to add the below code to `TransactionTable`
```
    @Override
    public MetricsReporter metricsReporter() {
      return BaseTransaction.this.reporter;
    }
```

File: core/src/main/java/org/apache/iceberg/SerializableTable.java [hunk 14]
```diff
@@ -83,6 +87,10 @@ protected SerializableTable(Table table) {
     this.encryption = table.encryption();
     this.locationProvider = table.locationProvider();
     this.refs = SerializableMap.copyOf(table.refs());
+    this.metricsReporterProperties =
+        table instanceof BaseTable
+            ? SerializableMap.copyOf(table.metricsReporter().properties())
+            : null;
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/SerializableTable.java:93]: we should rather do https://github.com/apache/iceberg/pull/7144#issuecomment-1504987582 than setting this to null here.

**Review by nastra — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/iceberg/SerializableTable.java (hunk 14 above)
Comment (wangyinsheng) [reply core/src/main/java/org/apache/iceberg/SerializableTable.java:93]: Sorry, I missed the comment when I try to fix the CI, I wll change it later.

**Comment by nastra** [comment]
@wangyinsheng I talked to @rdblue yesterday and we concluded that we don't want to carry over the catalog properties just for the metrics reporter. The better approach would be to make `MetricsReporter` extend `Serializable`. I opened https://github.com/wangyinsheng/iceberg/pull/1 against your branch to do that. Feel free to squash it into your changes

**Comment by wangyinsheng** [comment]
@nastra I am confused why should't carry over the catalog properties for metrics reporter? We already support to config metrics report through catalog properites.  And metrics reporter may contains non-serializable object, `KafakProducer` for example. We need properties to recreate non-serializable object after table been deserialized

**Comment by wangyinsheng** [comment]
please take a look @nastra @rdblue @aokolnychyi @danielcweeks

**Comment by wangyinsheng** [comment]
please take a look @nastra @rdblue @aokolnychyi  @danielcweeks

**Comment by nastra** [comment]
@wangyinsheng as mentioned previously, we still would like to make the `MetricsReporter` `Serializable`. For things that aren't `Serializable` in a metrics reporter, we would make them `transient` and then lazily initialize them

**Comment by rdblue** [comment]
@nastra, where are we with the fix for serialization here?

**Comment by nastra** [comment]
> @nastra, where are we with the fix for serialization here?

@rdblue I haven't heard back from @wangyinsheng after [my comment above ](https://github.com/apache/iceberg/pull/7144#issuecomment-1519866237), so I've opened https://github.com/apache/iceberg/pull/7370 a while ago for initial experimentation and then updated it to fix the serialization issue.

**Comment by wangyinsheng** [comment]
> > @nastra, where are we with the fix for serialization here?
> 
> @rdblue I haven't heard back from @wangyinsheng after [my comment above ](https://github.com/apache/iceberg/pull/7144#issuecomment-1519866237), so I've opened #7370 a while ago for initial experimentation and then updated it to fix the serialization issue.

@nastra @rdblue Sorry，I am working on other things for a while.  But I am still keep my opinion,  I think the `SerializableTable` is the best place to handle the `lazily initialization` stuff for `MetricsReporter`, we should not just make `MetricsReporter` serializable, which will throw the `lazily initialization` stuff to any implemention class of `MetricsReporter`
